DRILL-7763: Add Limit Pushdown to File Based Storage Plugins

cgivre opened the pull request:
# [DRILL-7763](https://issues.apache.org/jira/browse/DRILL-7763): Add Limit Pushdown to File Based Storage Plugins

## Description

As currently implemented, when querying a file, Drill will read the entire file even if a limit is specified in the query.  This PR does a few things:
* Refactors the `EasyGroupScan`, `EasySubScan`, and `EasyFormatConfig` to allow the option of pushing down limits.
* Applies this to all the EVF based format plugins which are: `Avro`, `LogRegex`, `PCAP`, `SPSS`, `Esri`, `Excel` and `Text (CSV)`. 
Due to JSON's fluid schema, it would be unwise to adopt the limit pushdown as it could result in very inconsistent schemata.

This capability is disabled by default, and has to be manually added to each format plugin if appropriate.

## Documentation
No user visible changes.  Queries with limits on large files are considerably faster.

## Testing
All existing unit tests are run.

vvysotskyi commented:
@cgivre, how it would work for the case when there was created multiple fragments with their own scan? From the code, it looks like every fragment would read the same number of rows specified in the limit. Also, will the limit operator be preserved in the plan if the scan supports limit pushdown?

Metastore also provides capabilities for pushing the limit, but it works slightly differently - it prunes files and leaves only minimum files number with specific row count. Would these two features coexist and work correctly?

KazydubB commented on contrib/format-spss/src/test/java/org/apache/drill/exec/store/spss/TestSpssReader.java:
```diff
@@ -108,7 +108,7 @@ public void testStarQuery() throws Exception {
   @Test
   public void testExplicitQuery() throws Exception {
-    String sql = "SELECT ID, Urban, Urban_value FROM dfs.`spss/testdata.sav` WHERE d16=4";
+    String sql = "SELECT ID, Urban, Urban_value FROM dfs.`spss/testdata.sav` WHERE d16=4 LIMIT 5";
```
Why not write another test method with `LIMIT` instead of editing the existing one without a `LIMIT`?
In case with `LIMIT`, should query plan be validated also?

cgivre replied:
@KazydubB 
I'll add additional unit tests this weekend.

cgivre replied:
Fixed.

sanel commented on contrib/format-esri/src/main/java/org/apache/drill/exec/store/esri/ShpBatchReader.java:
```diff
@@ -180,6 +182,10 @@ private String byteArrayToString(byte[] in) {
   }
 
   private void processShapefileSet(RowSetLoader rowWriter, final int gid, final Geometry geom, final Object[] dbfRow) {
+    if (maxRecords > 0 && rowWriter.rowCount() >= maxRecords) {
```
Same code happens to be used in multiple places? Maybe static function will make it clearer

cgivre replied:
Fixed in https://github.com/apache/drill/pull/2092/commits/90c82f22cf201c046964a8a8ca45d8d0f66b2168

cgivre commented:
@vvysotskyi 
Thanks for taking a look.  

> @cgivre, how it would work for the case when there was created multiple fragments with their own scan? From the code, it looks like every fragment would read the same number of rows specified in the limit. Also, will the limit operator be preserved in the plan if the scan supports limit pushdown?

Firstly, the format plugin has to explicitly enable the pushdown.  I don't have the best test infrastructure, so maybe you could assist with that, but I do believe that each fragment would read the same number of rows in their own scan.  Ideally, I'd like to fix that, but let's say you have 5 scans that are reading files with 1000 rows and you put a limit of 100 on the query.  Without this PR, my observation was that Drill will still read 5000 rows, whereas with this PR, it will only reduce that to 500.  

> 
> Metastore also provides capabilities for pushing the limit, but it works slightly differently - it prunes files and leaves only minimum files number with specific row count. Would these two features coexist and work correctly?

I didn't know about this feature in the metastore.  I would like for these features to coexist if possible.  Could you point me to some resources, or docs for this so that I can take a look?  Ideally, I'd like to make it such that we get the minimum files number from the metastore AND we get the row limit as well, so that we are looking at the absolute minimum amount of data.

For some background I was working on a project where I had several GB of PCAP files in multiple directories.  I found that Drill could query these files fairly rapidly, but it seemed to still have a lot of overhead in terms of how many files it was actually reading.  Separately, when I was working on the Splunk plugin (https://github.com/apache/drill/pull/2089), I discovered that virtually no storage plugins actually seemed to have a limit pushdown.  This was puzzling since the rules and logic for this were actually already in Drill and in the GroupScan.  On top of that, it's actually a fairly easy addition.  

Getting back to this PR, I wanted to see if it made a performance difference on querying some large files on my machine and the difference was shocking.  Simple queries and queries with a `WHERE` clause, which used to take seconds, would now be virtually instantaneous.  The difference is user experience is really shocking.  

Anyway, I'd appreciate any help you can give with respect to the metastore and incorporating that into the PR.

cgivre commented:
@vvysotskyi 
One more thought, it might make sense, if we can get the metastore sorted out, to do the same thing for parquet format.

vvysotskyi commented:
@cgivre, there already was a similar functionality for Parquet: https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/ParquetRecordReader.java#L53.
Please take a look at [`AbstractGroupScanWithMetadata.applyLimit()`](https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/AbstractGroupScanWithMetadata.java#L453) method - it contains required logic for pruning files. With your changes, you are overriding it, so it breaks this functionality. To coexist with this feature, please take a look at the implementation of this method for the parquet group scan, move common logic to AbstractGroupScanWithMetadata and use it in the easy group scan.

The behavior with the metastore usage is the following: if for example, we have 10 files with 100 records, and query with limit 5 is submitted, only the single file would be left in the group scan, since it contains all required records.

vvysotskyi commented on exec/java-exec/src/test/java/org/apache/drill/exec/metastore/TestMetastoreWithEasyFormatPlugin.java:
```diff
@@ -1079,14 +1079,14 @@ public void testFilesPruningWithLimit() throws Exception {
           .sql("select * from dfs.tmp.`%s` limit 1", tableName)
           .planMatcher()
-          .include("Limit", "numFiles=1,")
+          .include("Limit", "numFiles=12,")
```
This is the regression I mentioned in the previous comment. It is non-optimal to read all 12 files instead of the single one.

cgivre replied:
Reverted test to orginal

vvysotskyi commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/easy/EasyGroupScan.java:
```diff
@@ -172,6 +175,26 @@ private EasyGroupScan(final EasyGroupScan that) {
+  }
+
+  // Constructor to get the limit pushed down
+  private EasyGroupScan(final EasyGroupScan that, int maxRecords) {
```
Please do not introduce the new constructor. `maxRecords` is not final, so it is possible to assign its value after using the existing copy constructor.

cgivre replied:
Fixed

vvysotskyi commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/easy/EasyGroupScan.java:
```diff
@@ -17,31 +17,30 @@
-import java.util.ArrayList;
-import java.util.List;
-
+import com.fasterxml.jackson.annotation.JacksonInject;
```
Please keep the original imports order.

cgivre replied:
@vvysotskyi My IDE (IntelliJ) did this automatically.  Can we live with this?

cgivre commented:
@vvysotskyi I don't know if you're still looking at this or not, but I refactored the code and it's now ready for review.  I removed the `applyLimit()` function from the `EasyGroupScan` and made some minor modifications to the [`AbstractGroupScanWithMetadata.applyLimit()`](https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/AbstractGroupScanWithMetadata.java#L453) method so that it simply passes down the limit.  

At this point it is up to each format plugin to actually do something with it.  This should not affect the pruning at all.

cgivre commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/AbstractGroupScanWithMetadata.java:
```diff
@@ -451,19 +456,20 @@ public boolean supportsLimitPushdown() {
   @Override
   public GroupScan applyLimit(int maxRecords) {
-    maxRecords = Math.max(maxRecords, 1); // Make sure it request at least 1 row -> 1 file.
+    this.maxRecords = maxRecords;
```
Removed... the `TODO` is actually unnecessary.

cgivre commented on exec/java-exec/src/test/java/org/apache/drill/exec/metastore/TestMetastoreWithEasyFormatPlugin.java:
```diff
@@ -1079,15 +1079,17 @@ public void testFilesPruningWithLimit() throws Exception {
+          .include("Limit", "numFiles=3", "maxRecords=21")
           .match();
+      String plan = queryBuilder().sql("select * from dfs.tmp.`%s` limit 21", tableName).explainText();
+      System.out.println(plan);
```
Fixed.

cgivre commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/easy/EasyGroupScan.java:
```diff
@@ -387,7 +406,7 @@ public AnalyzeInfoProvider getAnalyzeInfoProvider() {
       // updates common row count and nulls counts for every column
       if (newScan.getTableMetadata() != null && files != null && newScan.getFilesMetadata().size() != files.size()) {
-        newScan.tableMetadata = TableMetadataUtils.updateRowCount(newScan.getTableMetadata(), files.values());
+        newScan.tableMetadata = TableMetadataUtils.updateRowCount(newScan.getTableMetadata(), files.values()); // TODO If limit is pushed down, adjust here...
```
Removed..  The `TODO` is actually unnecessary.

cgivre commented:
@vvysotskyi 
Thanks for looking at this.  I fixed the import order, removed the `TODO` and took care of the `print` statement.  

Regarding the one remaining comment about modifying the groupScan.  Originally, I had a separate constructor which applied the `maxLimit` to the GroupScan for this purpose.  I believe in previous comments, you asked to remove this, which I did. 
The goal here really is just make the limit available to `recordReaders` so that they can stop reading if it makes sense to do so.  If I shouldn't add an additional constructor to the `AbstractGroupScan`, and I shouldn't modify it, how would you suggest populating this field?

Thanks!

vvysotskyi commented:
@cgivre, under modifying the `groupScan`, I meant changing its state after calling `groupScan.applyLimit()` method, in this case, the `groupScan` shouldn't be modified, but in this method, a new group scan instance is created, so that instance should be updated only.

vvysotskyi commented:
Regarding import order, please also revert the change for other classes you have changed.

cgivre commented:
@vvysotskyi 
I fixed the import orders.  Regarding the extra constructor, I modified the `applyLimit` method a bit which I think does what you're asking.  Instead of modifying in place, it returns a new GroupScan with the limit value set.

vvysotskyi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/AbstractGroupScanWithMetadata.java:
```diff
@@ -454,26 +454,40 @@ public boolean supportsLimitPushdown() {
             tableRowCount, maxRecords);
+        // Return the group scan with the limit pushed down
+        if (this.maxRecords != maxRecords) {
+          this.setLimit(maxRecords);
```
This line modifies this instance of a group scan. Please create a new instance using a copy constructor, set the value there, and return it.

cgivre replied:
@vvysotskyi 
Thanks for your patience.  I think I got it this time.

vvysotskyi reviewed (commented):
@cgivre, the group scan is still modified in that method, I've added a comment where. It is important to avoid such modifications, since the planner caches rel nodes stats, but such changes make this cache invalid.

vvysotskyi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/AbstractGroupScanWithMetadata.java:
```diff
@@ -449,10 +454,22 @@ public boolean supportsLimitPushdown() {
     return true;
   }
 
+  public void setLimit (int limitToPushDown) {
```
This method is not used anymore. Could you please remove it?

cgivre replied:
@vvysotskyi 
Removed extra method.

vvysotskyi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/AbstractGroupScanWithMetadata.java:
```diff
@@ -453,6 +458,14 @@ public boolean supportsLimitPushdown() {
     GroupScanWithMetadataFilterer<?> prunedMetadata = getFilterer();
+
+    // Return the group scan with the limit pushed down
+    if (this.maxRecords != maxRecords) {
```
Placing this block here will also break files pruning logic. If the condition is true, pruning logic wouldn't be applied. Please move it to the places, where null is returned.

cgivre replied:
@vvysotskyi 
Done.  I did that originally, but it seemed like repetitive code to me.  Also, the pruning unit tests did pass with the logic at the top of the function.  However, I will defer to you on this and the pruning is quite complex and I don't want to break it.

vvysotskyi reviewed (approved):
@cgivre, thanks for making changes, +1

cgivre commented:
@vvysotskyi 
Thanks for the review!  I squashed commits.  I'll wait 24 hours before commiting.
